src/bridge/mod.rs

```diff
@@ -236,25 +252,52 @@ async fn launch(handler: NeovimHandler, grid_size: Option<Dimensions>) -> Result
     res.map(|()| session)
 }
 
-async fn run(session: NeovimSession) {
-    match session.io_handle.await {
-        Err(join_error) => error!("Error joining IO loop: '{}'", join_error),
-        Ok(Err(error)) => {
-            if !error.is_channel_closed() {
-                error!("Error: '{}'", error);
+async fn run(session: NeovimSession, proxy: EventLoopProxy<UserEvent>) {
+    let mut session = session;
+
+    if let Some(process) = session.neovim_process.as_mut() {
+        let neovim_exited = select! {
+            _ = &mut session.io_handle => false,
+            _ = process.wait() => {
+                log::info!("The Neovim process quit before the IO stream, waiting two seconds");
+                true
+            }
+        };
+
+        // We primarily wait for the stdio to finish, but due to bugs,
+        // for example, this one in in Neovim 0.9.5
+        // https://github.com/neovim/neovim/issues/26743
+        // it does not always finish.
+        // So wait for some additional time, both to make the bug obvious and to prevent incomplete
+        // data.
+        if neovim_exited {
+            let sleep = sleep(Duration::from_millis(2000));
+            tokio::pin!(sleep);
+            select! {
+                _ = session.io_handle => {}
+                _ = &mut sleep  => {
+                    log::info!("The IO stream was never closed, forcing Neovide to exit");
+                }
```

Comment:
Wouldn't `tokio::time::timeout()` be simpler here?

Reply:
Maybe, but it looks like the timeout consumes the future, and I want to re-use it after that.

Reply:
I changed it to a Tokio `timeout`